src/main/scala/chisel3/util/Mux.scala

```diff
@@ -63,10 +63,20 @@ object MuxLookup {
     * @return the value found or the default if not
     */
   def apply[S <: UInt, T <: Data] (key: S, default: T, mapping: Seq[(S, T)]): T = {
-    var res = default
-    for ((k, v) <- mapping.reverse)
-      res = Mux(k === key, v, res)
-    res
+    /* If the mapping is defined for all possible values of the key, then don't use the default value */
+    val (defaultx, mappingx) = try {
+      val k = mapping.map(_._1)
+      if ((math.pow(2, key.getWidth) == k.size) && (k.distinct.size == k.size)) {
```

Comment:
Due to Data equality being defined as referential equality, using `distinct.size` is not actually correct. There can also be non-literals.

As discussed in dev, we should instead check if all of the keys are literals and mutually exclusive (modulo select width*)

*I'm actually not sure that's correct, I believe when you compare larger things to smaller things it will sign/zero extend, so anything greater than the max value for the select can never match? We should test this.

Reply:
Let's say your key is 2-bit, and your lookup keys are `0.U`, `1.U`, `2.U` and `4096.U`. If you don't mask the lookup case keys before checking if their lit values are mutually exclusive, you see that there are four values, and conclude that it is an exhaustive case. However, since there is nothing that matches for `key === 3.U`, it is not actually an exhaustive case.

The false "phantom distinct value" is from `4096.U`, which would be correctly flagged as not distinct if the mapping keys' lit values are masked first.